server/src/db/catalog/partition.rs

```diff
@@ -231,16 +233,30 @@ impl Partition {
         self.tables.values().flat_map(|table| table.chunks.values())
     }
 
-    /// Return an iterator over chunks in this partition that
-    /// may pass the provided predicate
+    /// Return an iterator over chunks in this partition
+    ///
+    /// If `table_names` is Some(table_names) only returns chunks that store
+    /// data for a table in that list. Otherwise returns all chunks
     pub fn filtered_chunks<'a>(
         &'a self,
-        predicate: &'a Predicate,
+        table_names: Option<&'a BTreeSet<String>>,
     ) -> impl Iterator<Item = &Arc<RwLock<Chunk>>> + 'a {
         self.tables
             .iter()
-            .filter(move |(table_name, _)| predicate.should_include_table(table_name))
-            .flat_map(|(_, table)| table.chunks.values())
+            .filter_map(move |(partition_table_name, partition_table)| {
+                match table_names {
+                    Some(table_names) => {
+                        if table_names.contains(partition_table_name) {
+                            Some(partition_table.chunks.values())
+                        } else {
+                            None
+                        }
+                    }
+                    // No tablename provided
+                    None => Some(partition_table.chunks.values()),
+                }
```

Comment:
I'm a little concerned about the use of `Option<&BTreeSet<String>>` for `table_names` as opposed to `BTreeSet<String>`-- right now, passing `None` would have a different behavior than passing `Some(BTreeSet::new())`, and that would be surprising to me. 

This part could also be simpler if `table_names` wasn't an `Option`:

```suggestion
                if table_names.is_empty() || !table_names.contains(partition_table_name) {
                    None
                } else {
                    Some(partition_table.chunks.values())
                }
```

It *would* make all the cases where you're passing `None` a bit annoying as you'd have to make an empty set and pass a reference to that instead...

And then there's the question of whether `Predicate` should have `table_names: BTreeSet<String>,` instead as well, and I'm not sure of the implications/semantics of that.

What do you think?

Reply:
> I'm a little concerned about the use of Option<&BTreeSet<String>> for table_names as opposed to BTreeSet<String>-- right now, passing None would have a different behavior than passing Some(BTreeSet::new()), and that would be surprising to me.

I agree this is somewhat confusing. I think what is most confusing is `Some(empty BTreeSet)` isn't meaningful (it means return nothing). 

This is the same semantics of how `Predicate::table_names` works today, but in that case there is no way to construct the Some(empty BTreeSet).

The more I type this out, the more it seems like a good idea to add a specific enum for this like:

```rust
enum TableNameFilter {
  AllTables,
  NamedTables(BTreeSet<String>)
}
```

And then construct it accordingly. 

What do you think? It is semantically the same as `Option` but the different naming and construction may make it easier to understand

Reply:
Changed in 59fd67f6